sympy/solvers/ode/systems.py

```diff
@@ -0,0 +1,488 @@
+    or not.
+
+    Parameters
+    ==========
+
+    eqs: List
+        List of ODEs
+    funcs: List
+        List of dependent variables
+    t: Symbol
+        Independent variable of the equations in eqs
+
+    Returns
+    =======
+
+    match = {
+        'no_of_equation': len(eqs),
+        'eq': eqs,
+        'func': funcs,
+        'order': order,
+        'is_linear': is_linear,
+        'is_constant': is_constant,
+        'is_homogeneous': is_homogeneous,
+    }
+
+    Dict or None
+        Dict with values for keys:
+            1. no_of_equation: Number of equations
+            2. eq: The set of equations
+            3. func: List of dependent variables
+            4. order: A dictionary that gives the order of the
+                      dependent variable in eqs
+            5. is_linear: Boolean value indicating if the set of
+                          equations are linear or not.
+            6. is_constant: Boolean value indicating if the set of
+                          equations have constant coefficients or not.
+            7. is_homogeneous: Boolean value indicating if the set of
+                          equations are homogeneous or not.
+        This Dict is the answer returned if the eqs are linear and constant
+        coefficient. Otherwise, None is returned.
+    """
+    from sympy.solvers.solvers import solve
+
+    # Error for i == 0 can be added but isn't for now
+
+    # Removing the duplicates from the list of funcs
+    # meanwhile maintaining the order. This is done
+    # since the line in classify_sysode: list(set(funcs)
+    # cause some test cases to fail when gives different
+    # results in different versions of Python.
+    funcs = list(uniq(funcs))
+
+    # Check for len(funcs) == len(eqs)
+    if len(funcs) != len(eqs):
+        raise ValueError("Number of functions given is not equal to the number of equations %s" % funcs)
+
+    # ValueError when functions have more than one arguments
+    for func in funcs:
+        if len(func.args) != 1:
+            raise ValueError("dsolve() and classify_sysode() work with "
+            "functions of one variable only, not %s" % func)
+
+    # Getting the func_dict and order using the helper
+    # function
+    order = _get_func_order(eqs, funcs)
+
+    if not all(order[func] == 1 for func in funcs):
+        return None
+    else:
+
+        # TO be changed when this function is updated
+        system_order = 1
+
+    # Not adding the check if the len(func.args) for
+    # every func in funcs is 1
+
+    # Linearity check
+    try:
+        canon_eqs = solve(eqs, *[func.diff(t) for func in funcs])
+        canon_eqs = [func.diff(t) - canon_eqs[func.diff(t)] for func in funcs]
+        As, b = linear_ode_to_matrix(canon_eqs, funcs, t, system_order)
+
+    # When the system of ODEs is non-linear, either TypeError or
+    # a ValueError is raised. For both of these cases, None is returned.
+    except (ValueError, TypeError):
```

Comment:
If this is catching exceptions from `linear_ode_to_matrix` then the other lines should not be in the try block.

Why is it catching `ValueError` and `TypeError`? Maybe `linear_ode_to_matrix` should be made more consistent. I think that `linear_ode_to_matrix` should say which errors it raises in its docstring.

Reply:
@oscarbenjamin From what I can gather, it catches ```TypeError``` when we have non-linear ODEs, that is, since I am doing the computation for canonical euqation beforehand so that we would have to use the ```linear_ode_to_matrix``` function only once, sometimes it happens that the ```solve``` function doesn't give a ```dict``` type solution but a ```list of dicts``` type solution when there are multiple solutions for non-linear ODE terms.

Reply:
Exactly: catching TypeError like that hides bugs. In library code `solve` should be called with `dict=True` to make sure that the return type is consistent.

I don't like catching either ValueError or TypeError in general. We should have more specific exception classes (like `ODEOrderError`).

Reply:
We can raise a different type of error in ```linear_ode_to_matrix``` when we try to use the function ```linear_eq_to_matrix```. And then in the match function, we can catch that specific error and return```None```. Will that be alright?

Reply:
@oscarbenjamin I tried using ```solve``` with ```dict=True``` kwarg but it gave the same result as before.

Reply:
What result? With what input to solve/dsolve?

Reply:
```
In [28]:     x, y, z, w = symbols('x, y, z, w', cls=Function) 
    ...:     t = Symbol('t') 
    ...:     x1 = diff(x(t), t) 
    ...:     y1 = diff(y(t), t) 
    ...:     z1 = diff(z(t), t) 
    ...:     w1 = diff(w(t), t) 
    ...:     x2 = diff(x(t), t, t) 
    ...:     funcs = [x(t), y(t)] 
    ...:     funcs_2 = funcs + [z(t), w(t)]                                                                                                                                                                 

In [29]: eqs_2 = (5 * (x1**2) + 12 * x(t) - 6 * (y(t)), (2 * y1 - 11 * t * x(t) + 3 * y(t) + t))                                                                                                                                                                                                                                                                                         

In [31]: solve(eqs_2, *[func.diff(t) for func in funcs], dict=True)                                                                                                                                         
Out[31]: 
[{Derivative(x(t), t): -sqrt(-12*x(t)/5 + 6*y(t)/5),
  Derivative(y(t), t): 11*t*x(t)/2 - t/2 - 3*y(t)/2},
 {Derivative(x(t), t): sqrt(-12*x(t)/5 + 6*y(t)/5),
  Derivative(y(t), t): 11*t*x(t)/2 - t/2 - 3*y(t)/2}]
```

Reply:
That is the expected result of using `dict=True`: you should always get a list of dicts.

Reply:
@oscarbenjamin So, that is the reason of the ```TypeError```. I think it will be best to raise a new type of error like ```ODENonLinearError``` or something along those lines and raise this error whenever we encounter a non-linear ODE. We can either do that or we can just return the ```None```.

Reply:
I agree. I think that `linear_ode_to_matrix` could raise `ODENonlinearError`. This function can catch that exception and return None.

Reply:
Actually `linear_eq_to_matrix` should also be changed to raise a more specific exception like `NonlinearError`. Then both `linear_ode_to_matrix` and `linear_eq_to_matrix` can raise the same exception when the inputs are nonlinear.

We should document this in the docstrings for both functions using a "Raises" section.

Reply:
@oscarbenjamin So what we can do is: 1. We can create a new function ```canon_equations``` which can also raise ```ODENonlinearError```.
2. We add and raise ```NonLinearError``` in ```linear_eq_to_matrix```. 
3. We can catch ```NonLinearError``` in ```linear_ode_to_matrix``` and raise ```ODENonlinearError```.
4. We can catch ```ODENonlinearError``` in the match function and return ```None```.

Does this sound good or should the above approach but edited?

Reply:
I'm not sure I understand what `canon_equations` is but the other ideas all sound good.

We should also document the exceptions that are raised by `linear_eq_to_matrix` and `linear_ode_to_matrix` in the docstrings. Perhaps there could be a raises section:
```
Raises
=====

NonlinearError:
    if the system is not linear.
```
like from https://numpydoc.readthedocs.io/en/latest/format.html

Also there should be an example doctest of what happens if the ODEs are nonlinear.

Reply:
@oscarbenjamin I have started this work of adding errors. If we add these many error types, don't we have to add this in a special segment of the documentation listing all the error types?

Reply:
> I'm not sure I understand what `canon_equations` is but the other ideas all sound good.

```canon_equations``` will be a function that will bring a system of linear ODEs into canonical form. So, given a system as ```A1 X' + A2 X' = b1```, it will bring the system into a form: 
```
X' = A X + b
```
This will be useful for us when we implement weakly and strongly connected components logic.